content/blog/2022/2022-08-15-admin-operations-with-nodejs-launchdarkly.md

```diff
@@ -0,0 +1,699 @@
+
+```json
+{
+	"scripts": {
+    	"rateLimiter": "node rate_limiter.js"
+  	}
+}
+```
+
+Then we can execute the following command to run our app:
+
+```bash
+npm run rateLimiter
+```
+
+Note that using a feature flag for complex JSON configuration can be risky, because you don't get immediate feedback if the JSON you provided is valid or not (unless you look into the logs). So if you do this, be very careful not to provide invalid JSON in the feature flag variation.
+
+
+### Schedule Cron Jobs Dynamically
+
+Sometimes system admins need to schedule cron jobs to perform various tasks like gathering system metrics, generate reports, clearing or archiving logs, taking a backup, etc. Usually, these cron jobs are scheduled using `cronTime` expressions which are understood by cron executors. If there is a sudden need to change the `cronTime` expression of a particular job, then we can define it as a feature flag (whose value we can change on demand) and use it whenever the cron runs.
+
+For this, first we will run:
+
+```bash
+npm install cron
+```
+
+Then, we will define a variation in LaunchDarkly with string parameters which will take different `cronTime` expressions as values:
+
+{{% image alt="Rate Limit Error" src="images/posts/nodejs-backend-launchdarkly/Cron_Variation.png" %}}
+
+Next, we will define our cron job code which will retrieve the config from Launchdarkly and schedule the cron:
+
+```javascript
+import cron from 'cron';
+import LaunchDarkly from 'launchdarkly-node-server-sdk';
+
+const CronJob = cron.CronJob;
+const CronTime = cron.CronTime;
+
+// Initiating LaunchDarkly Client
+const LD_SDK_KEY = 'sdk-********-****-****-****-************';
+const userName = 'admin';
+const launchDarklyClient = LaunchDarkly.init( LD_SDK_KEY );
+
+launchDarklyClient.once('ready', async () => {    
+    const cronConfig = await launchDarklyClient.variation(
+      'cron-config',
+      {
+          key: userName
+      },
+      '*/4 * * * *' // Default fall-back variation value.
+    );
+
+    const job = new CronJob(cronConfig, function() {
+      run();
+    }, null, false)
+
+    let run = () => {
+      console.log('scheduled task called');
+    }
+
+    let scheduler = () => {
+      console.log('CRON JOB STARTED WILL RUN AS PER LAUNCHDARKLY CONFIG');
+      job.start();
+    }
+
+    let schedulerStop = () => {
+      job.stop();
+      console.log('scheduler stopped');
+    }
+
+    let schedulerStatus = () => {
+      console.log('cron status ---->>>', job.running);
+    }
+
+    let changeTime = (input) => {
+      job.setTime(new CronTime(input));
+      console.log('changed to every 1 second');
+    }
+
+    scheduler();
+    setTimeout(() => {schedulerStatus()}, 1000);
+    setTimeout(() => {schedulerStop()}, 9000);
+    setTimeout(() => {schedulerStatus()}, 10000);
+    setTimeout(() => {changeTime('* * * * * *')}, 11000);
+    setTimeout(() => {scheduler()}, 12000);
+    setTimeout(() => {schedulerStop()}, 16000);
+  }
+);
+```
```

Comment:
The above code snippet is not explained at all. Can you add a paragraph here that explains it shortly?

Reply:
I have added a small paragraph on what we are doing as part of this code segment. Please review.